src/compiler/checker.ts

```diff
@@ -8977,9 +8977,9 @@ namespace ts {
                 return getTypeForBindingElement(declaration as BindingElement);
             }
 
-            const isProperty = isPropertyDeclaration(declaration) || isPropertySignature(declaration);
+            const isProperty = isPropertyDeclaration(declaration) || isPropertySignature(declaration) || isJSDocPropertyTag(declaration);
             const isOptional = includeOptionality && (
-                isProperty && !!declaration.questionToken ||
+                (isPropertyDeclaration(declaration) || isPropertySignature(declaration)) && !!declaration.questionToken ||

```

Comment:
I don't understand why the definition of `isProperty` needs to be spilt in the `isOptional` check. Can you explain? Is it because there's no unified function that checks whether a property declaration, property signature or jsdoc property tag are optional?

Reply:
I think we need two different variables for `addOptionality` to work correctly

```ts
if (declaredType) {
    return addOptionality(declaredType, isProperty, isOptional);
}
```

`JSDocPropertyTag` does not have a `questionToken` property, so we cannot use `isProperty && !!declaration.questionToken` because TS reports about missing property in this case.

There is a `hasQuestionToken`, however, it handles all nodes that can have a `questionToken`, not sure if this is acceptable for this case.

Reply:
I was thinking of `isProperty && isOptionalDeclaration(declaration)`, where isOptionalDeclaration can give the right answer for all of property declarations, property signatures and jsdoc property tags.

Reply:
@sandersn I've moved this condition to utility function `isOptionalDeclaration`.